src/render/utf8.rs

```diff
@@ -0,0 +1,110 @@
+//! UTF-8 rendering, with 2 pixels per symbol.
+
+use crate::render::{Canvas as RenderCanvas, Pixel, Color};
+
+const CODEPAGE: [&str; 4] = [" ","\u{2584}","\u{2580}","\u{2588}"];
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum Unicode1x2 {
+    Dark, Light
+}
+
+impl Pixel for Unicode1x2 {
+    type Image = String;
+    type Canvas = Canvas;
+    fn default_color(color: Color) -> Unicode1x2 { color.select(Unicode1x2::Dark, Unicode1x2::Light) }
+    fn default_unit_size() -> (u32, u32) { (1, 1) }
+}
+
+impl Unicode1x2 {
+    fn value(&self) -> u8 {
+        match self {
+            Unicode1x2::Dark => {1}
+            Unicode1x2::Light => {0}
+        }
+    }
+    #[doc(hidden)]
+    fn parse_2_bits(sym: &u8) -> &'static str {
+        CODEPAGE[*sym as usize]
+    }
+}
+
+pub struct Canvas {
+    canvas: Vec<u8>,
+    width: u32,
+    dark_pixel: u8
+}
+
+impl RenderCanvas for Canvas {
+
+    type Pixel = Unicode1x2;
+    type Image = String;
+
+    
+    fn new(width: u32, height: u32, dark_pixel: Unicode1x2, light_pixel: Unicode1x2) -> Self { 
+        let a = vec![light_pixel.value(); (width * height) as usize];
```

Comment:
Add an `assert!` that `(width+1)/2 == height`

Reply:
But why? It works on non-rectangle canvas just fine :|

Reply:
In other non-text canvas, the `width` and `height` can be independently scaled and result in rectangular modules. This is not the case for 1x2 since the width:height ration is fixed.

Reply:
Oh, and also my implementation doesn't store it on per-symbol basis :|